File: Difficulty: Hard/Longest valid Parentheses/longest-valid-parentheses.cpp
//{ Driver Code Starts
// Initial Template for C++

#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
// User function Template for C++

class Solution{
public:
    int maxLength(string S){
        int l=0,r=0,maxlen=0;
        int n=S.size();
        
        for(int i=0;i<n;i++)
        {
            if(S[i]=='(')
            l++;
            
            else
            r++;
            
            if(l==r)
            maxlen=max(maxlen,r+l);//2, 4
            
            if(r>l)
            r=l=0;
        }
        r=l=0;
        
        for(int i=n-1;i>=0;i--)
        {
            if(S[i]=='(')
            l++;
            
            else
            r++;
            
            if(l==r)
            maxlen=max(maxlen,r+l);
            
            if(l>r)
            r=l=0;
        }
        return maxlen;
    }
};


//{ Driver Code Starts.

int main() {
    int t;
    cin >> t;
    while (t--) {
        string S;
        cin >> S;

        Solution ob;
        cout << ob.maxLength(S) << "\n";
    }
    return 0;
}
// } Driver Code Ends

File: Hard/Longest valid Parentheses/longest-valid-parentheses.java
//{ Driver Code Starts
//Initial Template for Java

import java.io.*;
import java.util.*;

class GFG{
    public static void main(String args[])throws IOException
    {
        BufferedReader in = new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(in.readLine());
        while(t-- > 0){
            String S = in.readLine();
            
            Solution ob = new Solution();
            System.out.println(ob.maxLength(S));
        }
    }
}
// } Driver Code Ends


//User function Template for Java
class Solution{
    static int maxLength(String s){
        // code here
      Stack<Integer> stack = new Stack<>();
      stack.push(-1);       // at first push -1 to stack
      int max=0;            //take max variable = 0 initially
      for(int i=0;i<s.length();i++){
      Character ch = s.charAt(i);
        if(ch == '('){
            stack.push(i);
        }
        else{
            stack.pop();
            if(stack.isEmpty()){
                stack.push(i);
            }
            max = Math.max(max,i-stack.peek());
        }
      }
    return max;
    
    }
}

